include/boost/math/tools/ieee754_linear.hpp

```diff
@@ -0,0 +1,371 @@
+//  (C) Copyright Ryan Elandt 2023.
+//  Use, modification and distribution are subject to the
+//  Boost Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_TOOLS_IEEE754_LINEAR_HPP
+#define BOOST_MATH_TOOLS_IEEE754_LINEAR_HPP
+
+#ifdef _MSC_VER
+#pragma once
+#endif
+
+#include <type_traits>
+
+
+#ifdef BOOST_HAS_FLOAT128
+#include <boost/multiprecision/float128.hpp>
```

Comment:
I think this breaks standalone mode as BOOST_HAS_FLOAR128 may be set but the multiprecision headers will not be available.  We also try to avoid mutual/cyclic dependencies like these.  What is the include actually used for?

Reply:
Good catch with the dependency. I've removed this include and replaced it with the following includes:
```c++
#include <boost/math/tools/config.hpp>  // For BOOST_MATH_FLOAT128_TYPE
#include <boost/cstdfloat.hpp>  // For numeric_limits support for 128 bit floats
```
There may be a better way of including the needed files that I'm unaware of.

Reply:
Unless you've changed something in that test, it uses nothing from multiprecision at all so the include should not be needed.

Reply:
As stated in a previous comment, I've removed the header `#include <boost/multiprecision/float128.hpp>` and replaced it with `#include <boost/cstdfloat.hpp>  // For numeric_limits support for 128 bit floats`. The `std_real_concept_check_128` test needs for `std::numeric_limits` specializations to be defined for the `BOOST_MATH_FLOAT128_TYPE`. If this header is removed, then test will **not** compile. (I tried this)

Why this occurs: the `std_real_concept_check_128` test makes the `std_real_concept` emulate a 128 bit floating point type. The tools in `ieee754_linear` recognize this as an IEEE 754 floating point type. The comment on line 133 goes on to say: 
```c++
// NOTE: returns Layout_IEEE754Linear<BOOST_MATH_FLOAT128_TYPE, ...> instead of Layout_IEEE754Linear<T, ...>
//       in order to work with non-trivial types that wrap trivial 128 bit floating point types.
```

Reply:
OK, I see the issue, but I think that we're looking at this the wrong way:

1) Not depending on boost/cstdfloat.hpp is a "good thing" because while that header is genuinely very useful, it's also rather fragile to changes in GCC releases.
2) I think the assumptions in `IsFloat128` (and maybe elsewhere) are wrong: just because a type has the correct size and properties, does NOT mean it's a trivial wrapper around a native type.  It could for example be a pointer (or two) to the actual implementation which just happens to have the same size.  mpfr_t is 24 bytes on my system so doesn't quite qualify, but if I'm reading it correctly, MPFR could be configured in a such a way that it was a 16 bye / 128 bit type which was set up to emulate a 128 bit float.  Another way of looking at this is that std_real_concept_type and the other concept checking types should be viewed as black boxes - their purpose is to emulate to some degree or another, some other random type about which we know nothing that the library may be instantiated with in the future.  You may be able to actually see their internals, but you really shouldn't look ;)

So I would change

```
   class IsFloat128 {
   public:
      template <typename T>
      static constexpr bool value() {
         return std::numeric_limits<T>::is_iec559 &&
                std::numeric_limits<T>::radix == 2 &&
                std::numeric_limits<T>::digits == 113 &&  // Mantissa has 112 bits + 1 implicit bit
                sizeof(T) == 16;
      }
   };
```

To maybe:

```
   class IsFloat128 {
   public:
      template <typename T>
      static constexpr bool value() {
         return std::numeric_limits<T>::is_iec559 &&
                std::numeric_limits<T>::radix == 2 &&
                std::numeric_limits<T>::digits == 113 &&  // Mantissa has 112 bits + 1 implicit bit
#ifdef BOOST_HAS_FLOAT128
                (std::is_floating_Point<T>::value || std::is_same<__float128, T>::value)
#else
                std::is_floating_Point<T>::value
#endif
      }
   };
```

Likewise for the other IsFloatXX classes.

Then change:

```
   template <typename T, typename U>
   class Layout_IEEE754Linear {
      static_assert(std::numeric_limits<T>::is_iec559, "Type must be IEEE 754 floating point.");
      static_assert(std::is_unsigned<U>::value, "U must be an unsigned integer type.");
      static_assert(sizeof(T) == sizeof(U), "Type and uint size must be the same.");

   public:
      using type_float = T;  // The linear floating point type
   };
```

To:

```
   template <typename T, typename U>
   class Layout_IEEE754Linear {
      static_assert(std::numeric_limits<T>::is_iec559 
#ifdef BOOST_HAS_FLOAT128
       || std::is_same<T, __float128>::value
#endif 
       , "Type must be IEEE 754 floating point.");
      static_assert(std::is_unsigned<U>::value, "U must be an unsigned integer type.");
      static_assert(sizeof(T) == sizeof(U), "Type and uint size must be the same.");

   public:
      using type_float = T;  // The linear floating point type
   };
```

Which would allow the header dependency to be removed as well.  I do appreciate it's less clean, but I suspect it's a safer option in the long run!

Reply:
> Not depending on boost/cstdfloat.hpp is a "good thing" because while that header is genuinely very useful, it's also rather fragile to changes in GCC releases.

Does the following change make things less fragile? Changing the include from:
```c++
#include <boost/cstdfloat.hpp>
```
to 
```c++
#if defined(BOOST_HAS_INT128) && defined(BOOST_HAS_FLOAT128)
#if !defined(BOOST_CSTDFLOAT_NO_LIBQUADMATH_LIMITS)
  #include <boost/math/cstdfloat/cstdfloat_limits.hpp>
#endif
#endif
```

>I think the assumptions in IsFloat128 (and maybe elsewhere) are wrong: just because a type has the correct size and properties, does NOT mean it's a trivial wrapper around a native type. It could for example be a pointer (or two) to the actual implementation which just happens to have the same size. [...] MPFR could be configured in a such a way [...] to emulate a 128 bit float. 

For the `value()` function of any of the `IsFloat` classes to return true, the following statement needs to be true:
```
std::numeric_limits<T>::is_iec559
```
The 32, 64, and 128 bit IEEE floats have a specific layout consisting of a sign bit, an exponent, and a fraction. A number type containing two pointers will not meet this standard. Skimming the MPFR codebase, this library is clearly set up with an awareness of IEEE 754 arithmetic. They don't appear to specialize numeric limits, and therefore don't set `std::numeric_limits<T>::is_iec559 = true`. I think the check that `std::numeric_limits<T>::is_iec559 = true` provides reasonable protection.

Reply:
Yup apologies, you're correct that `std::numeric_limits<T>::is_iec559` should provide a reasonable check given that that std defines a binary layout.

I would still prefer to avoid the cstdfloat includes if possible though - and we *know* that __float128 is an iec559 compatible type, so including it for that one case is overkill when we can just use `is_same<T, __float128>::value` ?

Other than that and one minor nit about <stdfloat> inclusion this looks good to go.

Reply:
I don't see a good way to do this. The code operates by duck typing a float type `T`. For example, something is determined to be a 32 bit binary floating point type based on the value of `IsFloat32::value<T>()`. Something is determined to be such a type if: 
```c++
std::numeric_limits<T>::is_iec559 &&
std::numeric_limits<T>::radix == 2 &&
std::numeric_limits<T>::digits == 24 &&  // Mantissa has 23 bits + 1 implicit bit
sizeof(T) == 4;
```
evaluates to true. If any of the four conditions above cannot be evaluated, then the code will fail to compile. The include is needed for it to compile, even if the method were to be unused.

----------

The alternative way of structuring this code is to manually match up binary floating point types with the appropriate methods, but this is messy and tends to miss valid use cases. `is_same<T, __float128>::value` might work on GCC, but what about an Intel compiler?

-----------

I'd like to avoid this include here too as it shouldn't be necessary. Perhaps the issue occurs here:
https://github.com/boostorg/math/blob/09d82da6e9884da86495279b331ff038dd1330b8/include/boost/math/tools/config.hpp#L399-L415
where a 128 bit floating point type is defined without defining `std::numeric_limits` for that type.